### ISSUE TITLE
apollo-link-state

### ISSUE THREAD

**Issue by lifeiscontent** [apollographql/eslint-plugin-graphql#99]
Now that apollo-link-state allows for local resolvers, this plugin throws errors on those local resolvers.

Can we get an update to pass the root resolver or something into the plugin to let eslint know about the extra information of the schema?

**Comment by adampash** [comment]
Alternately, even as a stop-gap if the plugin ignored fields with the @client directive, that'd be a nice start. (Though obviously supporting those would be nice in the long-run.)

**Comment by jnwng** [comment]
yea, i think having a client-side schema that can extend the schema provided would be interesting. let me poke around and see what people think about how that would work

in the meantime, @adampash's suggestion to provide an option to strip custom directives seems useful.

**Comment by loicplaire** [comment]
Does anyone have a work around until https://github.com/apollographql/eslint-plugin-graphql/pull/117 is merged?

**Comment by loicplaire** [comment]
If anyone else runs into that issue, my workaround until https://github.com/apollographql/eslint-plugin-graphql/pull/117 is merged is to redeclare the client directive in my apollo-link-state `typeDefs` e.g:

```js

**Comment by cantino** [comment]
@loicplaire, could you explain a bit more how to use this to prevent eslint-plugin-graphql from validating `@client` queries?

**Comment by donovanhiland** [comment]
In the meantime this seems to work:

```
gql`
  fragment Example {
    clientField @client ${
      /* eslint-disable-line graphql/template-strings */ ''
    }
  }
`
```

**Comment by nether-cat** [comment]
I've also realized that GraphQL type extensions are not supported by the eslint plugin. No matter where I define e.g. `extend type Payload { someFlag: Boolean }` or even `extend type Query { someField: String }`, the eslint plugin builds the schema correctly (if I put in some typo, I get corresponding errors) but when I use these fields, they are not accepted by the linting rules. (So far I have tried using `env: 'apollo'` as well as `env: 'literal'` in my config).

**Comment by jnwng** [comment]
> I've also realized that GraphQL type extensions are not supported by the eslint plugin. No matter where I define e.g. `extend type Payload { someFlag: Boolean }` or even `extend type Query { someField: String }`, the eslint plugin builds the schema correctly (if I put in some typo, I get corresponding errors) but whereever I use these fields, they are not accepted by the linting rules. (So far I have tried using `env: 'apollo'` as well as `env: 'literal'` in my config).

i dove deep on this and it has to do with the [`buildSchema` function we're using from `graphql-js`](https://github.com/apollographql/eslint-plugin-graphql/issues/227#issuecomment-493705087). i _think_ the resolution is likely to use the more extensive schema building functions in `graphql-tools`, and that sounds like it might be a solution here to extend the schema directives that are available instead of stripping them like #117.

somewhat relevant context, `graphql-js` [doesn't support custom directives](https://github.com/graphql/graphql-js/issues/41) right now, which is a large limitation for `eslint-plugin-graphql`

**Comment by sampsonjoliver** [comment]
Is this resolved?

**Comment by lifeiscontent** [comment]
No

**Comment by MrTibbles** [comment]
To overcome this use the [Apollo CLI download schema command](https://github.com/apollographql/apollo-tooling#apollo-clientdownload-schema-output) to generate the schema file to validate against. When generating the schema pass the `--includes` argument with a value of the file path to wher your local GQL type defs are, the ones that define the local state. Then also pass the `--endpoint` argument with a value of the URL for your GQL service:
```
apollo client:download-schema --includes \"./some/filepath/toLocalTypeDefs.ts\" --endpoint http://somehost/graphql
```

**Comment by damirbogd** [comment]
> To overcome this use the [Apollo CLI download schema command](https://github.com/apollographql/apollo-tooling#apollo-clientdownload-schema-output) to generate the schema file to validate against. When generating the schema pass the `--includes` argument with a value of the file path to wher your local GQL type defs are, the ones that define the local state. Then also pass the `--endpoint` argument with a value of the URL for your GQL service:
> 
> ```
> apollo client:download-schema --includes \"./some/filepath/toLocalTypeDefs.ts\" --endpoint http://somehost/graphql
> ```

@MrTibbles thank you so much, that's worked for me

```bash 
apollo client:download-schema -c './apollo.config.js' --includes './src/typeDefs.ts' 'graphql/schema.json'
```
.eslintrc.js config

```
rules: {
        'graphql/template-strings': [
            'error',
            {
                env: 'literal',
                schemaJson: require('./graphql/schema.json'),
            },
        ],
    },
```
apollo.config.js
```
module.exports = {
    client: {
        includes: [
            'src/**/*.gql',
            'src/typeDefs.ts',
            'src/vue-apollo.js',
            'src/**/*.vue',
        ],
        clientSchemaDirectives: ['client'],
        service: {
            name: 'my-service-name',
            localSchemaFile: './graphql/schema.graphql',
        },
    },
}
```
